prow/hook/server.go

```diff
@@ -17,238 +17,430 @@ limitations under the License.
+				s.Metrics.PluginHandleErrors.With(labels).Inc()
 			}
+			s.Metrics.PluginHandleDuration.With(labels).Observe(time.Since(start).Seconds())
+		}(p, h)
+	}
+	action := genericCommentAction(string(pr.Action))
+	if action == "" {
+		if !nonCommentPullRequestActions[pr.Action] {
+			l.Errorf(failedCommentCoerceFmt, "pull_request", string(pr.Action))
 		}
+		return
 	}
-	return matching
+	s.handleGenericComment(
+		l,
+		&github.GenericCommentEvent{
+			ID:           pr.PullRequest.ID,
+			GUID:         pr.GUID,
+			IsPR:         true,
+			Action:       action,
+			Body:         pr.PullRequest.Body,
+			HTMLURL:      pr.PullRequest.HTMLURL,
+			Number:       pr.PullRequest.Number,
+			Repo:         pr.Repo,
+			User:         pr.PullRequest.User,
+			IssueAuthor:  pr.PullRequest.User,
+			Assignees:    pr.PullRequest.Assignees,
+			IssueState:   pr.PullRequest.State,
+			IssueBody:    pr.PullRequest.Body,
+			IssueHTMLURL: pr.PullRequest.HTMLURL,
+		},
+	)
 }
 
-// demuxExternal dispatches the provided payload to the external plugins.
-func (s *Server) demuxExternal(l *logrus.Entry, externalPlugins []plugins.ExternalPlugin, payload []byte, h http.Header) {
-	h.Set("User-Agent", "ProwHook")
-	for _, p := range externalPlugins {
+func (s *Server) HandlePushEvent(l *logrus.Entry, pe github.PushEvent) {
+	l = l.WithFields(logrus.Fields{
+		github.OrgLogField:  pe.Repo.Owner.Name,
+		github.RepoLogField: pe.Repo.Name,
+		"ref":               pe.Ref,
+		"head":              pe.After,
+	})
+	l.Info("Push event.")
+	for p, h := range s.Plugins.PushEventHandlers(pe.Repo.Owner.Name, pe.Repo.Name) {
 		s.wg.Add(1)
-		go func(p plugins.ExternalPlugin) {
+		go func(p string, h plugins.PushEventHandler) {
 			defer s.wg.Done()
-			if err := s.dispatch(p.Endpoint, payload, h); err != nil {
-				l.WithError(err).WithField("external-plugin", p.Name).Error("Error dispatching event to external plugin.")
-			} else {
-				l.WithField("external-plugin", p.Name).Info("Dispatched event to external plugin")
+			agent := plugins.NewAgent(s.ConfigAgent, s.Plugins, s.ClientAgent, s.Metrics.Metrics, l, p)
+			start := time.Now()
+			labels := prometheus.Labels{"event_type": l.Data[eventTypeField].(string), "action": "none", "plugin": p}
+			if err := h(agent, pe); err != nil {
+				agent.Logger.WithError(err).Error("Error handling PushEvent.")
+				s.Metrics.PluginHandleErrors.With(labels).Inc()
 			}
-		}(p)
+			s.Metrics.PluginHandleDuration.With(labels).Observe(time.Since(start).Seconds())
+		}(p, h)
 	}
 }
 
-// dispatch creates a new request using the provided payload and headers
-// and dispatches the request to the provided endpoint.
-func (s *Server) dispatch(endpoint string, payload []byte, h http.Header) error {
-	req, err := http.NewRequest(http.MethodPost, endpoint, bytes.NewBuffer(payload))
-	if err != nil {
-		return err
+func (s *Server) HandleIssueEvent(l *logrus.Entry, i github.IssueEvent) {
+	l = l.WithFields(logrus.Fields{
```

Comment:
Shouldn't we set these log fields (and in all other `Handle` funcs) in the `GitHubEventServer`?

Reply:
Indeed.

Reply:
Addressed